superset-frontend/src/explore/components/controls/DateFilterControl/utils.ts

```diff
@@ -0,0 +1,209 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { SEPARATOR } from 'src/explore/dateFilterUtils';
+import {
+  CustomRangeDecodeType,
+  CustomRangeType,
+  DateTimeGrainType,
+  DateTimeModeType,
+} from './types';
+import { SEVEN_DAYS_AGO, MIDNIGHT } from './constants';
+
+/**
+ * RegExp to test a string for a full ISO 8601 Date
+ * Does not do any sort of date validation, only checks if the string is according to the ISO 8601 spec.
+ *  YYYY-MM-DDThh:mm:ss
+ *  YYYY-MM-DDThh:mm:ssTZD
+ *  YYYY-MM-DDThh:mm:ss.sTZD
+ * @see: https://www.w3.org/TR/NOTE-datetime
+ */
+const iso8601 = String.raw`\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(?:\.\d+)?(?:(?:[+-]\d\d:\d\d)|Z)?`;
+const datetimeConstant = String.raw`TODAY|NOW`;
+const grainValue = String.raw`[+-]?[1-9][0-9]*`;
+const grain = String.raw`YEAR|QUARTER|MONTH|WEEK|DAY|HOUR|MINUTE|SECOND`;
```

Comment:
Any reason to use normal case for these constants?

Reply:
Use UPPER CASE just for constant looks well. In fact, case insensitive in python datetime expression parser, in other words 'YEAR' == 'year' == 'YEAR' for backend datetime parser.